File: structy/Graphs/prereqs_possible.py
# Write a function, prereqs_possible, that takes in a number of courses (n) and prerequisites as arguments. 
# Courses have ids ranging from 0 through n - 1. A single prerequisite of (A, B) means that course A must be 
# taken before course B. The function should return a boolean indicating whether or not it is possible to complete all courses.

numCourses = 6
prereqs = [
  (0, 1),
  (2, 3),
  (0, 2),
  (1, 3),
  (4, 5),
]

def prereqs_possible(num_courses, prereqs):
    graph = build_graph(num_courses, prereqs)
    visited = set()
    
    for node in graph:
        if is_cycle(graph, node, set(), visited):
            return False
    return True

def is_cycle(graph, node, visiting, visited):
    if node in visited:
        return False

    if node in visiting:
        return True

    visiting.add(node)

    for neighbor in graph[node]:
        if is_cycle(graph, neighbor, visiting, visited):
            return True
    
    visited.add(node)
    visiting.remove(node)
    return False

def build_graph(num_courses, prereqs):
    graph = {}

    for course in range(num_courses):
        graph[course] = []

    for prereq in prereqs:
        a, b = prereq
        graph[a].append(b)

    return graph

print(prereqs_possible(numCourses, prereqs)) # -> True

File: structy/Graphs/prereqsPossible.js
function prereqsPossible(numCourses, prereqs) {
    const graph = buildGraph(numCourses, prereqs);
    const visited = new Set();

    for (const node in graph) {
        if (isCycle(graph, node, new Set(), visited) === true) return false;
    };

    return true;
};

function isCycle(graph, node, visiting, visited) {
    if (visited.has(node)) return false;
    if (visiting.has(node)) return true;

    visiting.add(node);

    for (const neighbor of graph[node]) {
        if (isCycle(graph, neighbor, visiting, visited) === true) return true
    };

    visiting.delete(node);
    visited.add(node);
    return false;
}

function buildGraph(numCourses, prereqs) {
    const graph = {};

    for (let i = 0; i < numCourses; i += 1) {
        graph[i] = []
    };

    for (const prereq of prereqs) {
        const [ a, b ] = prereq;

        graph[a].push(b)
    };

    return graph;
};


const numCourses = 6;
const prereqs = [
  [0, 1],
  [2, 3],
  [0, 2],
  [1, 3],
  [4, 5],
];
console.log(prereqsPossible(numCourses, prereqs)); // -> true
